sycl/doc/SYCL_compiler_and_runtime_design.md

```diff
@@ -42,6 +42,104 @@ back-end. Today middle-end transformations include just a couple of passes:
 - **Back-end** - produces native "device" code in ahead-of-time compilation
 mode.
 
+### SYCL support in Clang frontend
+
+SYCL support in Clang frontend can be split into the following components:
+
+- Device code outlining. Since SYCL is a single source programming model,
+the compiler has to separate device code from host code.
+
+- Lowering of lambda function objects and named function objects ("SYCL kernel
+functions"). To execute "SYCL kernel functions" on OpenCL devices some
+transformations are required.
+
+- Device code diagnostics.
+
+- Integration header generation.
+
+#### Device code outlining
+
+Here is a code example of a SYCL program that demonstrates compiler outlining
+work:
+
+```C++
+int foo(int x) { return ++x; }
+int bar(int x) { throw std::exception{"CPU code only!"}; }
+...
+using namespace cl::sycl;
+queue Q;
+buffer<int, 1> a{range<1>{1024}};
+Q.submit([&](handler& cgh) {
+      auto A = a.get_access<access::mode::write>(cgh);
+      cgh.parallel_for<init_a>(range<1>{1024}, [=](id<1> index) {
+        A[index] = index[0] * 2 + index[1] + foo(42);
+      });
+    }
+...
+```
+
+In this example, the SYCL compiler needs to compile the lambda expression passed
+to the `cl::sycl::handler::parallel_for` method, as well as the function `foo`
+called from the lambda expression for the device.
+The compiler must also ignore the `bar` function when we compile the
+"device" part of the single source code, as it's unused inside the device
+portion of the source code (the contents of the lambda expression passed to the
+`cl::sycl::handler::parallel_for` and any function called from this lambda
+expression).
+
+The current approach is to use the SYCL kernel attribute in the SYCL runtime to
+mark code passed to `cl::sycl::handler::parallel_for` as "kernel functions".
+The SYCL runtime library can't mark foo as "device" code - this is a compiler
+job: to traverse all symbols accessible from kernel functions and add them to
+the "device part" of the code marking them with the new SYCL device attribute.
+
```

Comment:
I think we should clarify the requirements we put on the function, which SYCL runtime can mark with `sycl_kernel` attribute.
I can remember at least these off the top of my head:
- Must be template function with at least two template parameters.
  - First parameter must represent "unique kernel name"
  - Second parameter must be the function object type
- Must have only one function parameter.
  - Parameter type must be second template parameter

```C++
  template <typename KernelName, typename KernelType/*, ...*/>
  __attribute__((sycl_kernel)) void sycl_kernel_function(KernelType KernelFuncObj) {
    KernelFuncObj();
  }
```

Other requirements can be extracted from the SemaSYCL.cpp. 
It's some informal API, which we should document somewhere.
Maybe it would be better to document in comments. What do you think?

Reply:
I think we can add some documentation to comments and this document both. But IMO requirements for function marked with 'sycl_kernel' attribute aren't related to the "device code oulining" feature but related to generation of kernel wrapper from this function. So I suggest adding it to the "SYCL kernel function object (functor or lambda) lowering" section.